### PR TITLE
[crypto] separate validator keys into consensus, staking, and network keys

### PR DESCRIPTION
**Pull request by punwai**
**Problem**
Due to optimization and compatibility reasons, we want to make different parts of the Sui network use different signature schemes. In summary, we want:
- Validators to use BLS12-381 for consensus/Fastpay
- Validators to use ed25519 or secp256k1 for securing TLS connections between validators
- Accounts to use ed25519 or secp256k1

This means that validators will now need to manage 3 keys. One for each function (note that validators must also own accounts, as they need a way to receive staking rewards). With this PR, validators now run on 3 keys rather than 1.

Must merge https://github.com/MystenLabs/private-testnet-genesis/pull/5 and https://github.com/MystenLabs/sui-operations/pull/296 after this PR in order to fix the production config files

### CODE REVIEW

File: crates/sui-types/src/crypto.rs [hunk 1]
```diff
@@ -143,6 +187,68 @@ impl AsRef<[u8]> for PublicKey {
...
+    }
+}
+
+impl From<Secp256k1PublicKey> for PublicKey {
+    fn from(key: Secp256k1PublicKey) -> Self {
+        PublicKey::Secp256k1KeyPair(key)
+    }
+}
+
+impl EncodeDecodeBase64 for PublicKey {
```
Comment (punwai) [inline crates/sui-types/src/crypto.rs:202]: Public keys now need to be serde'd to be stored in the validator configs.

File: crates/sui/src/genesis_ceremony.rs [hunk 2]
```diff
@@ -97,7 +101,9 @@ pub fn run(cmd: Ceremony) -> Result<()> {
 
         CeremonyCommand::AddValidator {
             name,
-            key_file,
+            validator_key_file,
+            staking_key_file: _,
```
Comment (joyqvq) [inline crates/sui/src/genesis_ceremony.rs:105]: is `_` used?

Comment (punwai) [reply crates/sui/src/genesis_ceremony.rs:105]: It is not yet used, we pinned it to the same key as validator_key_file

File: crates/sui-config/src/builder.rs [hunk 3]
```diff
@@ -93,12 +116,14 @@ impl<R: ::rand::RngCore + ::rand::CryptoRng> ConfigBuilder<R> {
             .map(|(i, validator)| {
                 let name = format!("validator-{i}");
                 let public_key: AuthorityPublicKeyBytes = validator.key_pair.public().into();
```
Comment (kchalkias) [inline crates/sui-config/src/builder.rs:118]: Ideally, in a future PR we have to rename `public_key` to something more meaningful (as we do for `network_key` and `account_key`)

File: crates/sui-types/src/crypto.rs [hunk 4]
```diff
@@ -145,6 +191,68 @@ impl AsRef<[u8]> for PublicKey {
...
+                    let pk = Ed25519PublicKey::from_bytes(&bytes[1..])?;
+                    Ok(PublicKey::Ed25519KeyPair(pk))
+                } else if x == &<Secp256k1PublicKey as SuiPublicKey>::SIGNATURE_SCHEME.flag() {
+                    let pk = Secp256k1PublicKey::from_bytes(&bytes[1..])?;
+                    Ok(PublicKey::Secp256k1KeyPair(pk))
+                } else {
+                    Err(eyre::eyre!("Invalid flag byte"))
+                }
+            }
+            _ => Err(eyre::eyre!("Invalid bytes")),
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:229]: In a future PR, we need to have a well defined tutorial on what do we need to change when adding a new flag (scheme).

File: crates/sui-types/src/crypto.rs [hunk 5]
```diff
@@ -153,14 +261,15 @@ impl PublicKey {
...
+
 //
 // Define Bytes representation of the Authority's PublicKey
 //
 
 #[serde_as]
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
 pub struct AuthorityPublicKeyBytes(
-    #[serde_as(as = "Readable<Base64, Bytes>")] [u8; AuthorityPublicKey::LENGTH],
+    #[serde_as(as = "Readable<Base64, Bytes>")] [u8; 1 + AuthorityPublicKey::LENGTH],
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:272]: For readability and consistency with how you do below, change to `AuthorityPublicKey::LENGTH + 1`

File: crates/sui-types/src/crypto.rs [hunk 6]
```diff
@@ -178,13 +287,15 @@ impl TryFrom<AuthorityPublicKeyBytes> for AuthorityPublicKey {
     type Error = signature::Error;
 
     fn try_from(bytes: AuthorityPublicKeyBytes) -> Result<AuthorityPublicKey, Self::Error> {
-        AuthorityPublicKey::from_bytes(bytes.as_ref()).map_err(|_| Self::Error::new())
+        AuthorityPublicKey::from_bytes(&bytes.as_ref()[1..]).map_err(|_| Self::Error::new())
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:290]: Do we need to check that the flag byte is the expected?

**Review by kchalkias — APPROVED** [review]
<3

**Pull request merged**